File: run_comprehensive_analysis.py
#!/usr/bin/env python3
"""
Main Comprehensive Analysis Workflow Script
Hosyana Review - Decision Support Methods Analysis

This script orchestrates the complete analysis workflow using the organized file structure.
It runs all components in the correct order and generates all outputs.

Usage:
    python3 run_comprehensive_analysis.py

Outputs:
    - data/analysis_results/ : CSV files with analysis results
    - data/reports/ : JSON reports and logs
    - figures/sankey_plots/ : Publication-ready PDF Sankey plots
    - figures/interactive/ : Interactive HTML visualizations
    - docs/ : Documentation and summaries
"""

import os
import subprocess
import sys
import time
from pathlib import Path


def print_header(title):
    """Print a formatted header"""
    print("\n" + "=" * 80)
    print(f" {title}")
    print("=" * 80)


def print_step(step, description):
    """Print a formatted step"""
    print(f"\n🔄 Step {step}: {description}")
    print("-" * 60)


def run_command(command, description):
    """Run a command and handle errors"""
    print(f"Running: {description}")
    print(f"Command: {command}")

    try:
        result = subprocess.run(
            command, shell=True, check=True, capture_output=True, text=True, cwd="."
        )

        if result.stdout:
            print(
                "Output:",
                result.stdout[:500] + ("..." if len(result.stdout) > 500 else ""),
            )

        print("✅ Completed successfully")
        return True

    except subprocess.CalledProcessError as e:
        print(f"❌ Error: {e}")
        if e.stdout:
            print(f"stdout: {e.stdout}")
        if e.stderr:
            print(f"stderr: {e.stderr}")
        return False


def check_file_exists(filepath, description):
    """Check if a file exists and print status"""
    if os.path.exists(filepath):
        size = os.path.getsize(filepath) / 1024  # KB
        print(f"✅ {description}: {filepath} ({size:.1f} KB)")
        return True
    else:
        print(f"❌ Missing: {description}: {filepath}")
        return False


def main():
    """Main workflow execution"""

    print_header("COMPREHENSIVE HOSYANA REVIEW ANALYSIS WORKFLOW")
    print("Processing ALL bibliography files with organized output structure")
    print(f"Working directory: {os.getcwd()}")

    # Verify we're in the right directory
    if not os.path.exists("hosyana_review.Rproj"):
        print("❌ Error: Not in hosyana_review directory")
        print("Please run this script from the hosyana_review project root")
        sys.exit(1)

    # Check required directories exist
    required_dirs = [
        "bib/bib_raw",
        "scripts/python",
        "scripts/r",
        "data/analysis_results",
        "data/reports",
        "figures/sankey_plots",
        "figures/interactive",
        "docs",
    ]

    print("\n📁 Verifying directory structure...")
    for directory in required_dirs:
        if os.path.exists(directory):
            print(f"✅ {directory}")
        else:
            print(f"❌ Missing: {directory}")
            os.makedirs(directory, exist_ok=True)
            print(f"   Created: {directory}")

    # Count bibliography files
    bib_files = list(Path("bib/bib_raw").glob("*.bib"))
    print(f"\n📚 Found {len(bib_files)} bibliography files to process")

    start_time = time.time()

    # Step 1: Run comprehensive Python analysis
    print_step(1, "Python Comprehensive Analysis")
    success = run_command(
        "python3 scripts/python/final_comprehensive_analysis.py",
        "Analyzing all bibliography files and detecting methods",
    )

    if not success:
        print("❌ Failed to complete Python analysis. Stopping workflow.")
        sys.exit(1)

    # Step 2: Generate interactive Python visualizations
    print_step(2, "Interactive Visualizations")
    success = run_command(
        "python3 scripts/python/create_sankey_plots.py",
        "Creating interactive HTML Sankey plots and visualizations",
    )

    if not success:
        print("⚠️  Warning: Interactive visualizations failed, continuing...")

    # Step 3: Generate R-compatible CSV
    print_step(3, "R-Compatible Data Generation")
    success = run_command(
        "Rscript scripts/r/create_comprehensive_csv.R",
        "Converting Python results to R-compatible format",
    )

    if not success:
        print("⚠️  Warning: R CSV generation failed, continuing...")

    # Step 4: Generate R-style Sankey plots
    print_step(4, "Publication-Ready Sankey Plots")
    success = run_command(
        "Rscript scripts/r/create_comprehensive_sankey.R",
        "Creating publication-ready PDF Sankey plots",
    )

    if not success:
        print("⚠️  Warning: R Sankey plots failed, continuing...")

    # Step 5: Generate comprehensive report
    print_step(5, "HTML Report Generation")
    success = run_command(
        "Rscript -e \"rmarkdown::render('index.Rmd')\"",
        "Generating comprehensive HTML report",
    )

    if not success:
        print("⚠️  Warning: HTML report generation failed, continuing...")

    # Verify outputs
    print_header("OUTPUT VERIFICATION")

    # Check main analysis files
    print("\n📊 Analysis Results:")
    check_file_exists(
        "data/analysis_results/FINAL_methods_analysis.csv", "Main analysis CSV"
    )
    check_file_exists(
        "data/analysis_results/COMPREHENSIVE_methods_classification.csv",
        "R-compatible CSV",
    )

    # Check reports
    print("\n📋 Reports:")
    check_file_exists("data/reports/FINAL_analysis_report.json", "Analysis summary")
    check_file_exists("data/reports/FINAL_sankey_data.json", "Sankey data")
    check_file_exists("data/reports/final_analysis.log", "Processing log")

    # Check visualizations
    print("\n📈 Visualizations:")
    sankey_files = [
        "figures/sankey_plots/comprehensive_sankey_1970.pdf",
        "figures/sankey_plots/comprehensive_sankey_1980.pdf",
        "figures/sankey_plots/comprehensive_sankey_1990.pdf",
        "figures/sankey_plots/comprehensive_sankey_2000.pdf",
        "figures/sankey_plots/comprehensive_sankey_2010.pdf",
        "figures/sankey_plots/comprehensive_sankey_2020.pdf",
        "figures/sankey_plots/comprehensive_sankey_2024.pdf",
    ]

    sankey_count = 0
    for sankey_file in sankey_files:
        if check_file_exists(sankey_file, f"Sankey plot"):
            sankey_count += 1

    interactive_files = [
        "figures/interactive/decade_method_sankey.html",
        "figures/interactive/method_evolution_sankey.html",
        "figures/interactive/methods_by_decade_bar.html",
        "figures/interactive/method_trends_lines.html",
    ]

    interactive_count = 0
    for interactive_file in interactive_files:
        if check_file_exists(interactive_file, "Interactive visualization"):
            interactive_count += 1

    # Check documentation
    print("\n📚 Documentation:")
    check_file_exists("docs/COMPREHENSIVE_ANALYSIS_SUMMARY.md", "Analysis summary")
    check_file_exists("index.html", "Main HTML report")

    # Final summary
    elapsed_time = time.time() - start_time
    print_header("WORKFLOW COMPLETE")

    print(f"⏱️  Total execution time: {elapsed_time / 60:.1f} minutes")
    print(f"📚 Bibliography files processed: {len(bib_files)}")
    print(f"📊 PDF Sankey plots created: {sankey_count}/7")
    print(f"📈 Interactive visualizations: {interactive_count}/4")

    # Quick statistics from main analysis file
    if os.path.exists("data/analysis_results/FINAL_methods_analysis.csv"):
        try:
            import pandas as pd

            df = pd.read_csv("data/analysis_results/FINAL_methods_analysis.csv")
            papers_with_methods = len(df[df["HAS_METHODS"] == 1])
            detection_rate = papers_with_methods / len(df) * 100

            print(f"📄 Total papers analyzed: {len(df):,}")
            print(f"🔍 Papers with methods detected: {papers_with_methods:,}")
            print(f"📈 Detection rate: {detection_rate:.1f}%")

        except ImportError:
            print("📄 Analysis files created (pandas not available for stats)")
        except Exception as e:
            print(f"📄 Analysis files created (could not read stats: {e})")

    print("\n🎉 COMPREHENSIVE ANALYSIS WORKFLOW COMPLETED SUCCESSFULLY!")
    print("\nNext steps:")
    print("  • Review results in data/analysis_results/")
    print("  • View PDF plots in figures/sankey_plots/")
    print("  • Open interactive visualizations in figures/interactive/")
    print("  • Read analysis summary in docs/COMPREHENSIVE_ANALYSIS_SUMMARY.md")
    print("  • View complete report: open index.html in browser")


if __name__ == "__main__":
    main()

File: run_comprehensive_analysis.R
#!/usr/bin/env Rscript
# Main Comprehensive Analysis Workflow Script (R)
# Hosyana Review - Decision Support Methods Analysis
#
# This script orchestrates the complete R analysis workflow using the organized file structure.
# It runs all R components and generates outputs in the correct directories.
#
# Usage:
#     Rscript run_comprehensive_analysis.R
#
# Outputs:
#     - data/analysis_results/ : R-compatible CSV files
#     - figures/sankey_plots/ : Publication-ready PDF Sankey plots
#     - index.html : Comprehensive HTML report

# Load required packages
suppressPackageStartupMessages({
  library(knitr)
  library(rmarkdown)
})

# Function to print formatted headers
print_header <- function(title) {
  cat("\n", paste(rep("=", 80), collapse = ""), "\n")
  cat(" ", title, "\n")
  cat(paste(rep("=", 80), collapse = ""), "\n\n")
}

# Function to print formatted steps
print_step <- function(step, description) {
  cat("\n🔄 Step", step, ":", description, "\n")
  cat(paste(rep("-", 60), collapse = ""), "\n")
}

# Function to run R scripts and handle errors
run_r_script <- function(script_path, description) {
  cat("Running:", description, "\n")
  cat("Script:", script_path, "\n")

  tryCatch(
    {
      source(script_path)
      cat("✅ Completed successfully\n")
      return(TRUE)
    },
    error = function(e) {
      cat("❌ Error:", e$message, "\n")
      return(FALSE)
    }
  )
}

# Function to check if file exists
check_file_exists <- function(filepath, description) {
  if (file.exists(filepath)) {
    size <- round(file.info(filepath)$size / 1024, 1) # KB
    cat("✅", description, ":", filepath, "(", size, "KB)\n")
    return(TRUE)
  } else {
    cat("❌ Missing:", description, ":", filepath, "\n")
    return(FALSE)
  }
}

# Main workflow function
main <- function() {
  print_header("COMPREHENSIVE HOSYANA REVIEW ANALYSIS WORKFLOW (R)")
  cat("Processing data with organized R workflow\n")
  cat("Working directory:", getwd(), "\n")

  # Verify we're in the right directory
  if (!file.exists("hosyana_review.Rproj")) {
    cat("❌ Error: Not in hosyana_review directory\n")
    cat("Please run this script from the hosyana_review project root\n")
    quit(status = 1)
  }

  # Check required directories exist
  required_dirs <- c(
    "data/analysis_results",
    "data/reports",
    "figures/sankey_plots",
    "scripts/r",
    "docs"
  )

  cat("\n📁 Verifying directory structure...\n")
  for (directory in required_dirs) {
    if (dir.exists(directory)) {
      cat("✅", directory, "\n")
    } else {
      cat("❌ Missing:", directory, "\n")
      dir.create(directory, recursive = TRUE, showWarnings = FALSE)
      cat("   Created:", directory, "\n")
    }
  }

  # Check if Python analysis results exist
  python_results <- "data/analysis_results/FINAL_methods_analysis.csv"
  if (!file.exists(python_results)) {
    cat("❌ Error: Python analysis results not found:", python_results, "\n")
    cat("Please run the Python analysis first:\n")
    cat("   python3 run_comprehensive_analysis.py\n")
    quit(status = 1)
  }

  start_time <- Sys.time()

  # Step 1: Generate R-compatible CSV
  print_step(1, "R-Compatible Data Generation")
  success <- run_r_script(
    "scripts/r/create_comprehensive_csv.R",
    "Converting Python results to R-compatible format"
  )

  if (!success) {
    cat("❌ Failed to create R-compatible CSV. Stopping workflow.\n")
    quit(status = 1)
  }

  # Step 2: Generate R-style Sankey plots
  print_step(2, "Publication-Ready Sankey Plots")
  success <- run_r_script(
    "scripts/r/create_comprehensive_sankey.R",
    "Creating publication-ready PDF Sankey plots"
  )

  if (!success) {
    cat("⚠️  Warning: R Sankey plots failed, continuing...\n")
  }

  # Step 3: Generate comprehensive R Markdown report
  print_step(3, "HTML Report Generation")
  tryCatch(
    {
      cat("Rendering index.Rmd to HTML...\n")
      rmarkdown::render("index.Rmd")
      cat("✅ HTML report generated successfully\n")
    },
    error = function(e) {
      cat("⚠️  Warning: HTML report generation failed:", e$message, "\n")
    }
  )

  # Verify outputs
  print_header("OUTPUT VERIFICATION")

  # Check analysis files
  cat("\n📊 Analysis Results:\n")
  check_file_exists(
    "data/analysis_results/FINAL_methods_analysis.csv",
    "Python analysis CSV"
  )
  check_file_exists(
    "data/analysis_results/COMPREHENSIVE_methods_classification.csv",
    "R-compatible CSV"
  )

  # Check Sankey plots
  cat("\n📈 Sankey Plots:\n")
  sankey_files <- c(
    "figures/sankey_plots/comprehensive_sankey_1970.pdf",
    "figures/sankey_plots/comprehensive_sankey_1980.pdf",
    "figures/sankey_plots/comprehensive_sankey_1990.pdf",
    "figures/sankey_plots/comprehensive_sankey_2000.pdf",
    "figures/sankey_plots/comprehensive_sankey_2010.pdf",
    "figures/sankey_plots/comprehensive_sankey_2020.pdf",
    "figures/sankey_plots/comprehensive_sankey_2024.pdf"
  )

  sankey_count <- 0
  for (sankey_file in sankey_files) {
    if (check_file_exists(sankey_file, "Sankey plot")) {
      sankey_count <- sankey_count + 1
    }
  }

  # Check documentation
  cat("\n📚 Documentation:\n")
  check_file_exists("docs/COMPREHENSIVE_ANALYSIS_SUMMARY.md", "Analysis summary")
  check_file_exists("index.html", "Main HTML report")

  # Final summary
  elapsed_time <- as.numeric(difftime(Sys.time(), start_time, units = "mins"))
  print_header("R WORKFLOW COMPLETE")

  cat("⏱️  Total execution time:", round(elapsed_time, 1), "minutes\n")
  cat("📊 PDF Sankey plots created:", sankey_count, "/7\n")

  # Quick statistics if CSV available
  if (file.exists("data/analysis_results/COMPREHENSIVE_methods_classification.csv")) {
    tryCatch(
      {
        df <- read.csv("data/analysis_results/COMPREHENSIVE_methods_classification.csv")
        cat("📄 Papers with methods in R CSV:", nrow(df), "\n")

        # Get year range
        if ("year" %in% names(df)) {
          year_range <- range(df$year, na.rm = TRUE)
          cat("📅 Year coverage:", year_range[1], "-", year_range[2], "\n")
        }
      },
      error = function(e) {
        cat("📄 R CSV created (could not read stats):", e$message, "\n")
      }
    )
  }

  cat("\n🎉 R ANALYSIS WORKFLOW COMPLETED SUCCESSFULLY!\n")
  cat("\nR-specific outputs:\n")
  cat("  • R-compatible data: data/analysis_results/COMPREHENSIVE_methods_classification.csv\n")
  cat("  • Publication plots: figures/sankey_plots/comprehensive_sankey_*.pdf\n")
  cat("  • HTML report: index.html\n")
  cat("\nFor complete workflow including Python analysis, run:\n")
  cat("  python3 run_comprehensive_analysis.py\n")
}

# Execute main function if script is run directly
if (!interactive()) {
  main()
}
